### PR TITLE
Add a simple implementation of CanvasRenderingContext2d.fillText

### LINKED ISSUE #11681: Implement CanvasRenderingContext2d.fillText
Spec: https://html.spec.whatwg.org/multipage/scripting.html#dom-context-2d-filltext
Code: `components/script/dom/canvasrenderingcontext2d.rs`, `components/script/dom/weibidls/CanvasRenderingContext2D.webidl`, `components/canvas/canvas_paint_thread.rs`

(I hope I identified correctly the code files - followed the style in #5411)

### PR DESCRIPTION
**Pull request by kaiakz**
I added a simple implementation of CanvasRenderingContext2d.fillText.
Some code are merged from @mikrut, and I fixed a bug about text scaling.
Also, the bug of text rotation should be fixed after `raqote` merged my other PR.

---

- [X] `./mach build -d` does not report any errors
- [X] `./mach test-tidy` does not report any errors
- [X] These changes fix (part of) #11681 (GitHub issue number if applicable)

- [X] There are tests for these changes

### CODE REVIEW

**Review by jdm — APPROVED** [review]

**Comment by jdm** [comment]
```
  ▶ FAIL [expected PASS] /2dcontext/text-styles/canvas_text_font_001.htm
  │   → /2dcontext/text-styles/canvas_text_font_001.htm 54ed3fe6baf2d8bc9bddcdb0e82725a5c568a6e6
  └   → /2dcontext/text-styles/canvas_text_font_001-ref.htm 652917f3c91833309587adc905f58a56e6998530
```
This means that you will need to run:
1. `./mach test-wpt /2dcontext/text-styles/canvas_text_font_001.htm --log-raw wpt`
1. `./mach update-wpt`
1. `git add tests/wpt/metadata/2dcontext`
And commit the changes.

**Comment by kaiakz** [comment]
@jdm 
`./mach test-wpt /2dcontext/text-styles/canvas_text_font_001.htm --log-raw wpt` says: 
``
web-platform-test
~~~~~~~~~~~~~~~~~
Ran 1 checks (1 tests)
Expected results: 0
Unexpected results: 1
  test: 1 (1 fail)

Unexpected Results
------------------
FAIL /2dcontext/text-styles/canvas_text_font_001.htm - /2dcontext/text-styles/canvas_text_font_001.htm 1fa1aef7b66b7af2d04621d380782299ec9c0e10
/2dcontext/text-styles/canvas_text_font_001-ref.htm d8c11dca7773f5977e977513c7310bd60d853355
 0:14.12 INFO Closing logging queue
 0:14.12 INFO queue closed
``
How can i fix it? Does it means I have to implement the `canvas_text_font`?

**Comment by jdm** [comment]
@kaiakz I gave you the steps to fix it - the current implementation test fails because the reference rendering does not match the test rendering, because the font support isn't implemented. The steps I suggested update the expected test result to indicate that the test will not pass, so the test failure won't be reported as a problem.

**Comment by jdm** [comment]
r? @pcwalton 
Could you check if this looks like a reasonable first cut at a canvas text implementation, disregarding the lack of font selection?

File: components/canvas/raqote_backend.rs
Comment (pcwalton) [inline components/canvas/raqote_backend.rs]: This should not sandbox-safe. It should be done in the chrome process, not the content process.

Comment (pcwalton) [inline components/canvas/raqote_backend.rs]: Needs a TODO because this should eventually use the same infrastructure as layout (i.e. layout should be updated to use font-kit as well).

Comment (pcwalton) [reply components/canvas/raqote_backend.rs]: Also, this is in the chrome process, right? It needs to not be in the content process because the content process shouldn't have access to the `SystemSource`. (It's too dangerous; on macOS for example it needs a connection to the font server, which has had a lot of security vulnerabilities.)

Comment (jdm) [reply components/canvas/raqote_backend.rs]: Yes, the canvas code runs in the chrome process.

Comment (kaiakz) [reply components/canvas/raqote_backend.rs]: So, I need to do:

1. add a TODO to `raqote_backend.rs`
2. move `font_kit` code from `raqote_backend.rs` to `canvas_data.rs` and pass the `SystemSource` things through` the parameter?

 

> Also, this is in the chrome process, right? It needs to not be in the content process because the content process shouldn't have access to the `SystemSource`.

I noticed there is a `FontCacheThread`.

Comment (jdm) [reply components/canvas/raqote_backend.rs]: Let's keep the font_kit code in raqote_backend.rs for now until it's more clear how the proper font integration will work. The FontCacheThread is integrated with many other parts of the system, so converting that code to use font_kit will be part of resolving that TODO.

Comment (kaiakz) [reply components/canvas/raqote_backend.rs]: @jdm 
If I understood well, I moved all font_kit code to `raqote_backend.rs`, including some functions: 
 `replace_whitespace` & `get_text_width`( They are parts of `text preparation algorithm`, in fact, I just moved the algorithm to `raqote_backend.rs`).

**Comment by jdm** [comment]
@bors-servo r+
Thanks!

**Comment by jdm** [comment]
Looks like upgrading raqote brings in a new version of sw-composite which does not work on ARM right now:
```
error[E0432]: unresolved import `std::arch::x86_64`
   --> /root/.cargo/registry/src/github.com-1ecc6299db9ec823/sw-composite-0.7.5/src/lib.rs:690:20
    |
690 |     use std::arch::x86_64::{
    |                    ^^^^^^ could not find `x86_64` in `arch`
error[E0432]: unresolved import `std::arch::x86_64`
   --> /root/.cargo/registry/src/github.com-1ecc6299db9ec823/sw-composite-0.7.5/src/lib.rs:719:16
    |
719 | use std::arch::x86_64::__m128i;
    |                ^^^^^^ could not find `x86_64` in `arch`
error[E0432]: unresolved import `std::arch::x86_64`
   --> /root/.cargo/registry/src/github.com-1ecc6299db9ec823/sw-composite-0.7.5/src/lib.rs:724:20
    |
724 |     use std::arch::x86_64::{
    |                    ^^^^^^ could not find `x86_64` in `arch`
error: aborting due to 3 previous errors
For more information about this error, try `rustc --explain E0432`.
error: could not compile `sw-composite`.
```
To fix this, we need to run `cargo update -p sw-composite` to pull in https://github.com/jrmuizel/sw-composite/commit/d071cc90befb28fb6471c33b2e6664c6927e77f6.

**Comment by jrmuizel** [comment]
If you upgrade to raqote 0.7.13 this issue should be fixed now.

**Comment by kaiakz** [comment]
@jdm @jrmuizel So what I need to do is run `cargo update -p sw-composite` or `cargo update -p raqote`?

**Comment by jdm** [comment]
Upgrading raqote will work.

**Comment by jdm** [comment]
@jrmuizel The raqote/sw-composite update seems to have triggered a bunch of new test passes, as well as some new panics:
```
  ▶ TIMEOUT [expected OK] /2dcontext/shadows/2d.shadow.pattern.alpha.html
  │ 
  │ 
  │ 
  │ 
  │ index 18446744073709546816 out of range for slice of length 5000 (thread CanvasThread, at src/libcore/slice/mod.rs:2674)
  │ stack backtrace:
  │    0: servo::backtrace::print
  │    1: servo::main::{{closure}}
  │    2: std::panicking::rust_panic_with_hook
  │              at src/libstd/panicking.rs:474
  │    3: rust_begin_unwind
  │              at src/libstd/panicking.rs:378
  │    4: core::panicking::panic_fmt
  │              at src/libcore/panicking.rs:85
  │    5: core::slice::slice_index_len_fail
  │              at src/libcore/slice/mod.rs:2674
  │    6: <raqote::blitter::ImageRepeatAlphaShader as raqote::blitter::Shader>::shade_span
  │    7: <raqote::blitter::ShaderMaskBlitter as raqote::blitter::Blitter>::blit_span
  │    8: raqote::draw_target::DrawTarget::composite
  │    9: raqote::draw_target::DrawTarget::fill
  │   10: canvas::raqote_backend::<impl canvas::canvas_data::GenericDrawTarget for raqote::draw_target::DrawTarget>::fill
  │   11: canvas::raqote_backend::<impl canvas::canvas_data::GenericDrawTarget for raqote::draw_target::DrawTarget>::fill_rect
  │   12: canvas::canvas_data::CanvasData::fill_rect
  │   13: canvas::canvas_paint_thread::CanvasPaintThread::process_canvas_2d_message
  │   14: std::sys_common::backtrace::__rust_begin_short_backtrace
  │   15: _ZN3std9panicking3try7do_call17hf57d287875f14debE.llvm.9274590977348817245
  │   16: __rust_maybe_catch_panic
  │              at src/libpanic_unwind/lib.rs:86
  │   17: core::ops::function::FnOnce::call_once{{vtable.shim}}
  │   18: <alloc::boxed::Box<F> as core::ops::function::FnOnce<A>>::call_once
  │              at /rustc/5d04ce67fd14538d03fa47a2598f80d49fd564c6/src/liballoc/boxed.rs:1017
  │   19: <alloc::boxed::Box<F> as core::ops::function::FnOnce<A>>::call_once
  │              at /rustc/5d04ce67fd14538d03fa47a2598f80d49fd564c6/src/liballoc/boxed.rs:1017
  │       std::sys_common::thread::start_thread
  │              at src/libstd/sys_common/thread.rs:13
  │       std::sys::unix::thread::Thread::new::thread_start
  │              at src/libstd/sys/unix/thread.rs:80
  │   20: start_thread
  │   21: __clone
  └ [2020-02-25T21:37:37Z ERROR servo] index 18446744073709546816 out of range for slice of length 5000
```

**Comment by jrmuizel** [comment]
The cause of the panic is probably: https://github.com/jrmuizel/raqote/commit/66cb86ea1ef07c8e81a88d252bc6e9ad888919ac

**Comment by jrmuizel** [comment]
If someone wants to put together a standalone raqote test case I'm sure it will be easy for me to fix the issue.

**Comment by kaiakz** [comment]
Before I rebase the code, how can I fix it?

**Comment by pylbrecht** [comment]
> If someone wants to put together a standalone raqote test case I'm sure it will be easy for me to fix the issue.

If no one's calling dibs on this, I'll try to reproduce and write a failing test case for it.

**Comment by kaiakz** [comment]
@pylbrecht 
OK:smile:
Is there anythings I can help you?

**Comment by pylbrecht** [comment]
> @pylbrecht
> That's fine 😄
> Is there anythings I can help you? I would like to learn how to do this

Of course, that's even better!

In general what I would do is translating the [test case's behavior](https://github.com/web-platform-tests/wpt/blob/master/2dcontext/shadows/2d.shadow.pattern.alpha.html) into a separate raqote example project.
This is how I would approach this:
1. Create a new raqote project (`cargo new --bin raqote-example`, don't forget to add raqote as a dependency to `Cargo.toml`). This is where I would try to reproduce the panic documented here: https://github.com/servo/servo/pull/25782#issuecomment-591091630. This way I wouldn't have to rebuild Servo all the time, which takes a lot more time than building the small example project.
1. The call stack of https://github.com/servo/servo/pull/25782#issuecomment-591091630 tells me that the panic occurs somewhere after `9: raqote::draw_target::DrawTarget::fill`. So this is the call I would choose to reproduce the panic in the raqote example project.
1. Next I would try to figure out what arguments have been passed to `fill()` that resulted in the panic, so I can reconstruct the call in the raqote example project. The two options that pop into my head for this are either adding `println!("{:?}", arg)`s to the corresponding Servo code (`raqote_backend.rs::GenericDrawTarget::fill()` IIRC) or running the test with a debugger (like gdb) and inspecting the call to `fill()` there:
`./mach test-wpt tests/wpt/web-platform-tests/2dcontext/shadows/2d.shadow.pattern.alpha.html --debug-build --debugger=gdb`
1. Here's a sketch of how the example code could look like:
```rust
extern crate raqote;
use raqote::*;

fn main() {
    let mut dt = DrawTarget::new(300, 300);
    // ctx.fillStyle = '#f00';
    let red = Source::Solid(SolidSource::from_unpremultiplied_argb(255, 255, 0, 0));
    // ctx.fillRect(0, 0, 100, 50);
    dt.fill_rect(0., 0., 100., 50., &red, &DrawOptions::new());

    // we can ignore these two calls
    // ctx.shadowOffsetY = 50;
    // ctx.shadowColor = '#00f';

    // var pattern = ctx.createPattern(document.getElementById('transparent50.png'), 'repeat');
    // ctx.fillStyle = pattern;
    // this needs to contain `transparent50.png`'s image data from the test case https://github.com/web-platform-tests/wpt/blob/master/2dcontext/shadows/2d.shadow.pattern.alpha.html
    // the image can be found here: https://github.com/web-platform-tests/wpt/blob/master/images/transparent50.png
    let image = Image { width: .., height: .., data: .. };
    let image_source = Source::Image(image, ..);

    // ctx.fillRect(0, -50, 100, 50);
    let mut pb = PathBuilder::new();
    pb.rect(0., -50., 100., 50.);
    dt.fill(&pb.finish(), &image_source, &DrawOptions::new());
}
```
For further reference I recommend looking at [raqote's docs](https://docs.rs/raqote/0.7.13/raqote/) and feel free to ask me any questions, either in the comments here or [chat.mozilla.org](https://chat.mozilla.org).

**Comment by pylbrecht** [comment]
I just accidentally ran into the same panic with this code:
```rust
extern crate raqote;
extern crate euclid;

use raqote::*;
use euclid::*;

fn main() {
    let mut dt = DrawTarget::new(300, 300);
    let options = DrawOptions::new();
    let green = Source::Solid(SolidSource::from_unpremultiplied_argb(255, 0, 255, 0));

    dt.fill_rect(0., 0., 100., 50., &green, &options);

    let img = Image {
        width: 100,
        height: 50,
        data: &vec![0xffff0000; 100*50],
    };

    let identity = Transform::identity();

    let img_src = Source::Image(img, ExtendMode::Repeat, FilterMode::Nearest, identity);
    let transform = dt.get_transform();
    dt.set_transform(&transform.pre_translate(vec2(50., 25.)));
    dt.fill_rect(-50., -25., 100., 50., &img_src, &options);

    dt.fill_rect(0., 0., 50., 25., &green, &options);

    dt.write_png("example.png");
}
```
Maybe it has something to to with `ExtendMode::Repeat` and `fill_rect()` having negative arguments?
```rust
    let img_src = Source::Image(img, ExtendMode::Repeat, FilterMode::Nearest, identity);
    dt.fill_rect(-50., -25., 100., 50., &img_src, &options);
```

**Comment by jdm** [comment]
Thanks for that @pylbrecht! I've filed https://github.com/jrmuizel/raqote/issues/116 describing the source of the issue.

**Comment by jdm** [comment]
If you rebase and then run `cargo update -p raqote` and commit the changes, we should be able to merge this :)

**Comment by kaiakz** [comment]
@jdm 
I have finished the job, so let's try it agsin. :smile:

**Comment by jdm** [comment]
```
  ▶ Unexpected subtest result in /offscreen-canvas/line-styles/2d.line.cap.closed.html:
  └ PASS [expected FAIL] Line caps are not drawn at the corners of an unclosed rectangle
  ▶ Unexpected subtest result in /offscreen-canvas/fill-and-stroke-styles/2d.gradient.radial.inside3.worker.html:
  └ PASS [expected FAIL] 2d
  ▶ Unexpected subtest result in /offscreen-canvas/line-styles/2d.line.cap.closed.worker.html:
  └ PASS [expected FAIL] Line caps are not drawn at the corners of an unclosed rectangle
  ▶ Unexpected subtest result in /offscreen-canvas/fill-and-stroke-styles/2d.gradient.radial.inside3.html:
  └ PASS [expected FAIL] OffscreenCanvas test: 2d.gradient.radial.inside3
  ▶ Unexpected subtest result in /2dcontext/line-styles/2d.line.cap.closed.html:
  └ PASS [expected FAIL] Line caps are not drawn at the corners of an unclosed rectangle
  ▶ Unexpected subtest result in /2dcontext/fill-and-stroke-styles/2d.gradient.radial.inside3.html:
  └ PASS [expected FAIL] Canvas test: 2d.gradient.radial.inside3
```
Just need to update test expectations for these tests now!

**Comment by kaiakz** [comment]
@jdm 
I have updated the wpt expectations: offscreen-canvas && 2dcontext, but there still have fails in other wpt: fetch, html, css... Should I ignore them?

**Comment by pylbrecht** [comment]
> but there still have fails in other wpt: fetch, html, css... Should I ignore them?

Those failures could be intermittent failures. We will see in the results of the CI pipeline I just triggered.

**Comment by pylbrecht** [comment]
@bors-servo retry

```
1 unexpected results that are NOT known-intermittents:
  ▶ TIMEOUT [expected OK] /html/infrastructure/urls/dynamic-changes-to-base-urls/historical.sub.xhtml
  │ 
  │ 
  │ 
  │ 
  │ 0:00:00.628103771 24550 0x7fa36c0394f0 FIXME                default gstutils.c:3981:gst_pad_create_stream_id_internal:<appsrc0:src> Creating random stream-id, consider implementing a deterministic way of creating a stream-id
  │ 0:00:00.628791173 24550 0x7fa36c0394f0 WARN                typefind gsttypefindelement.c:996:gst_type_find_element_chain_do_typefinding:<typefindelement0> error: Stream doesn't contain enough data.
  │ 0:00:00.641786022 24550 0x7fa36c0394f0 WARN                typefind gsttypefindelement.c:996:gst_type_find_element_chain_do_typefinding:<typefindelement0> error: Can't typefind stream
  │ 0:00:00.642016878 24550 0x55f9f5f79280 ERROR             gst-player gstplayer.c:1175:error_cb:<player0> ERROR: from element /GstPlayBin:playbin/GstURIDecodeBin:uridecodebin0/GstTypeFindElement:typefindelement0: Stream doesn't contain enough data.
  │ 
  │ 0:00:00.642037372 24550 0x55f9f5f79280 ERROR             gst-player gstplayer.c:1177:error_cb:<player0> Additional debug info:
  │ gsttypefindelement.c(996): gst_type_find_element_chain_do_typefinding (): /GstPlayBin:playbin/GstURIDecodeBin:uridecodebin0/GstTypeFindElement:typefindelement0:
  │ Can't typefind stream
  │ 
  │ 0:00:00.642059726 24550 0x55f9f5f79280 ERROR             gst-player gstplayer.c:1052:emit_error:<player0> Error: Error from element /GstPlayBin:playbin/GstURIDecodeBin:uridecodebin0/GstTypeFindElement:typefindelement0: Could not determine type of stream.
  │ Stream doesn't contain enough data.
  │ gsttypefindelement.c(996): gst_type_find_element_chain_do_typefinding (): /GstPlayBin:playbin/GstURIDecodeBin:uridecodebin0/GstTypeFindElement:typefindelement0:
  │ Can't typefind stream (gst-player-error-quark, 0)
  │ 0:00:00.667473101 24550 0x7fa36c0394f0 FIXME                default gstutils.c:3981:gst_pad_create_stream_id_internal:<appsrc1:src> Creating random stream-id, consider implementing a deterministic way of creating a stream-id
  │ 0:00:00.671900244 24550 0x7fa36c0394f0 WARN                typefind gsttypefindelement.c:996:gst_type_find_element_chain_do_typefinding:<typefindelement1> error: Stream doesn't contain enough data.
  │ 0:00:00.671922074 24550 0x7fa36c0394f0 WARN                typefind gsttypefindelement.c:996:gst_type_find_element_chain_do_typefinding:<typefindelement1> error: Can't typefind stream
  └ [2020-03-19T10:44:05Z ERROR script::dom::htmlmediaelement] Player error: "Error from element /GstPlayBin:playbin/GstURIDecodeBin:uridecodebin0/GstTypeFindElement:typefindelement0: Could not determine type of stream.\nStream doesn\'t contain enough data.\ngsttypefindelement.c(996): gst_type_find_element_chain_do_typefinding (): /GstPlayBin:playbin/GstURIDecodeBin:uridecodebin0/GstTypeFindElement:typefindelement0:\nCan\'t typefind stream"
```

Might be related to https://github.com/servo/servo/issues/23343

**Comment by pylbrecht** [comment]
@kaiakz, looks like you got all relevant tests covered. Let's wait for @jdm's approval.

**Comment by pylbrecht** [comment]
The layout-2020 equivalents must be updated as well (see https://github.com/servo/servo/pull/25940#issuecomment-598841120 for an explanation why). The layout-2020 tests' metadata files are located in `tests/wpt/metadata-layout-2020/`.
```
1 unexpected results that are NOT known-intermittents:
  ▶ Unexpected subtest result in /2dcontext/line-styles/2d.line.cap.closed.html:
  └ PASS [expected FAIL] Line caps are not drawn at the corners of an unclosed rectangle
```
```
2 unexpected results that are NOT known-intermittents:
  ▶ Unexpected subtest result in /2dcontext/fill-and-stroke-styles/2d.gradient.interpolate.zerosize.fillText.html:
  │ FAIL [expected PASS] Canvas test: 2d.gradient.interpolate.zerosize.fillText
  │   → assert_equals: d[12824] expected 0 but got 28
  │ 
  │ testColor@http://web-platform.test:8000/2dcontext/resources/canvas-tests.js:78:22
  │ _assertGreen@http://web-platform.test:8000/2dcontext/resources/canvas-tests.js:84:22
  │ @http://web-platform.test:8000/2dcontext/fill-and-stroke-styles/2d.gradient.interpolate.zerosize.fillText.html:31:13
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1958:25
  │ _addTest/</<@http://web-platform.test:8000/2dcontext/resources/canvas-tests.js:66:15
  │ Test.prototype.step@http://web-platform.test:8000/resources/testharness.js:1958:25
  └ _addTest/<@http://web-platform.test:8000/2dcontext/resources/canvas-tests.js:63:11
  ▶ Unexpected subtest result in /2dcontext/fill-and-stroke-styles/2d.gradient.radial.inside3.html:
  └ PASS [expected FAIL] Canvas test: 2d.gradient.radial.inside3
```

**Comment by kaiakz** [comment]
@pylbrecht 
How did you update the layout-2020 tests' metadata files? `./mach test-wpt` & `./mach update-wpt` do not work for me.

**Comment by jdm** [comment]
Sorry, let's just wait for #25960 to merge.

**Comment by pylbrecht** [comment]
I did it manually: copying the corresponding .ini file from the regular `tests/wpt/metadata/` directory to the `tests/wpt/metadata-layout-2020/` directory in case of a FAIL, or deleting the corresponding .ini file in case of a PASS.
But I think you also can wait until https://github.com/servo/servo/pull/25960 is merged.

**Comment by jdm** [comment]
@bors-servo retry
* #23290

**Comment by paulrouget** [comment]
This makes servo crash on Windows: https://github.com/servo/servo/issues/26015

**Comment by paulrouget** [comment]
unwrap: https://github.com/servo/font-kit/blob/3d3b79ef9b644a49aa723cb3af0a32d882a3fdf9/src/source.rs#L146

**Pull request merged**
